### PR TITLE
[Merged by Bors] - feat: add deprecation script

### PR DESCRIPTION
**Pull request by adomani**
The script contained in this file tries to automatically generate deprecations for declarations that were renamed in a PR.

The script is entirely text-based, so it makes several simplifying assumptions and may be easily confused.

#### Assumptions

The script works under the assumption that the only modifications between `master` and the current branch are renames of lemmas that should be deprecated.

The script inspects the relevant `git diff`, extracts the old name and the new one and uses this information to insert deprecated aliases as needed.

Most other differences may confuse the script, although there is some slack.

Please, do not try to push the boundaries of the script, since it is quite simple-minded!

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: scripts/add_deprecations.sh [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+##  for each modified declaration in `<file>`.
+##  The separators `@@@` delimit different declarations.
+##  The separators `||||` are later replaced by line breaks.
+## To use a specific date, replace $(date +%Y-%m-%d) with 2024-04-17 for instance
+mkDeclAndDepr () {
+  git diff --unified=0 "${commit}" "${1}" |
+    awk -v regex="${begs}" -v date="$(date +%Y-%m-%d)" '
+    function depr(ol,ne) {
+      line=sprintf("@[deprecated (since := \"%s\")]||||alias %s := %s", date, ol, ne)
+      if(length(line) <= 102) { sub(/\|\|\|\|/, " ", line) }
```
Comment (adomani) [inline scripts/add_deprecations.sh:49]:
```suggestion
      if(length(line) <= 103) { sub(/\|\|\|\|/, " ", line) }
```

**Comment by mattrobball** [comment]
This worked well for me with #18191

**Comment by adomani** [comment]
I'm glad to hear that the script works, also on macOS.  It is fairly simple-minded, but if it gets used more extensively, I can look into making it more robust.  The declarations_diff script started in a similar way and seems to perform quite well.

**Comment by mattrobball** [comment]
> @mattrobball I see that there is a [deprecations commit](https://github.com/leanprover-community/mathlib4/pull/18191/commits/e802e12cbd94ddf5ddf3965c50f8e9020c7c9d09) in your PR: was that autogenerated? Did you have to do any post-editing to it?

Yes and no

**Comment by adomani** [comment]
Great, that's awesome as it also tested the automatic line-breaking part that is new.

**Comment by TwoFX** [comment]
When I tried out this script on #18288, it missed the lemmas `ι_preservesColimitsIso_hom` and `ι_preservesColimitsIso_inv`, presumably because the names start with a greek letter. Not sure whether fixing this would fall under "do not try to push the boundaries of the script".

**Comment by TwoFX** [comment]
> If the actual regex of starts of identifiers is available

The relevant code is [here](https://github.com/leanprover/lean4/blob/f54bce2abbbd7a379b0a1ad505d178384baad37d/src/Init/Meta.lean#L77-L101). So I guess a regex for the first letter of an identifier could be

```
[a-zA-Z_\u3b1-\u3ba\u3bc-\u3c9\u391-\u39f\u3a1-\u3a2\u3a4-\u3a9\u3ca-\u3fb\u1f00-\u1ffe\u2100-\u214f\u1d49c-\u1d59f]
```

**Comment by adomani** [comment]
Awesome, thank you very much!  I added your regex to the script.

If it turns out that there are other issues with this, I may switch to a "negative" regex: simply excluding some characters that certainly do not begin an identifier (e.g. parentheses, digits, punctuation marks...) and maybe generate deprecations for commands that were not creating a declaration.

**Comment by YaelDillies** [comment]
There are still some issues with the script: I tried using it in #18330 and here is the result: https://github.com/leanprover-community/mathlib4/pull/18330/commits/d455e51043e3831db5e58499661068e5bbfb1161. You see that it is for example adding `@[deprecated (since := "2024-10-28")] alias  := isInducing` in `Topology.DenseEmbedding`

**Comment by adomani** [comment]
Yaël, thank you, this is very useful!  I will try to debug this, but note that I did not hard-code non-printing of "nameless" deprecations, so I consider the whole run in your file a great success!  :smile:

**Comment by adomani** [comment]
Ah, the lemma that the script is using is *new*!  There is no deprecation to be made, right?

**Comment by YaelDillies** [comment]
[There are still some issues with the scripts duplicating aliases](https://github.com/leanprover-community/mathlib4/pull/18398/commits/07b1b6356ddd436192ff678bed8291fca97df5b6). The pattern is that I rename a declaration from `A` to `B`, then change declaration `C` (but do not rename it). The script results in
```
lemma B := ...

@[deprecated] alias A := B

lemma C := ...

@[deprecated] alias A := C
```
It's as if the script "forgot" it already added a deprecation for `A`

**Comment by adomani** [comment]
Yaël, I should have fixed this bug: I knew what the fix was when I read your description, but then it took me ages to find a test that would show a difference!  In your heuristic, `lemma C` should be completely new, not a modification of an existing one.  Anyway, the new script should handle this situation now.

I also added a further feature: the script report to stderr a summary of all the deprecations that it applied.  This is a hand-rolled `git diff`, but it might be convenient to have the summary.

**Comment by YaelDillies** [comment]
Amazing! I don't have any other big refactor waiting on which I could test this further, so

maintainer merge

**Comment by kim-em** [comment]
@adomani would you mind waiting until #14539 lands, then merging master and updating `scripts/README.md`?

bors d+

**Comment by adomani** [comment]
bors merge

**Pull request closed without merging**